src/Shared/IInternable.cs

```diff
@@ -0,0 +1,346 @@
+    }
+
+    /// <summary>
+    /// A wrapper over char[].
+    /// </summary>
+    internal struct CharArrayInternTarget : IInternable
+    {
+        /// <summary>
+        /// Start index for the string
+        /// </summary>
+        private readonly int _startIndex;
+
+        /// <summary>
+        /// The held array
+        /// </summary>
+        private readonly char[] _target;
+
+        /// <summary>
+        /// Pointless comment about constructor.
+        /// </summary>
+        internal CharArrayInternTarget(char[] target, int count)
+            : this(target, 0, count)
+        {
+        }
+
+        /// <summary>
+        /// Pointless comment about constructor.
+        /// </summary>
+        internal CharArrayInternTarget(char[] target, int startIndex, int count)
+        {
+#if DEBUG
+            if (startIndex + count > target.Length)
+            {
+                ErrorUtilities.ThrowInternalError("wrong length");
+            }
+#endif
+            _target = target;
+            _startIndex = startIndex;
+            Length = count;
+        }
+
+        /// <summary>
+        /// The length of the target.
+        /// </summary>
+        public int Length { get; }
+
+        /// <summary>
+        /// Indexer into the target. Presumed to be fast.
+        /// </summary>
+        public char this[int index]
+        {
+            get
+            {
+                if (index > _startIndex + Length - 1 || index < 0)
+                {
+                    ErrorUtilities.ThrowInternalError("past end");
+                }
+
+                return _target[index + _startIndex];
+            }
+        }
+
+        /// <summary>
+        /// Convert target to string. Presumed to be slow (and will be called just once).
+        /// </summary>
+        public bool ReferenceEquals(string other)
+        {
+            return false;
+        }
+
+        /// <summary>
+        /// Convert target to string. Presumed to be slow (and will be called just once).
+        /// </summary>
+        public string ExpensiveConvertToString()
+        {
+            // PERF NOTE: This will be an allocation hot-spot because the char[] is finally determined to
+            // not be internable. There is still only one conversion of char[] into string it has just
+            // moved into this single spot.
+            return new string(_target, _startIndex, Length);
+        }
+
+        /// <summary>
+        /// Compare target to string. Assumes string is of equal or smaller length than target.
+        /// </summary>
+        public bool StartsWithStringByOrdinalComparison(string other)
+        {
+#if DEBUG
+            ErrorUtilities.VerifyThrow(other.Length <= Length, "should be at most as long as target");
+#endif
+            // Backwards because the end of the string is (by observation of Australian Government build) more likely to be different earlier in the loop.
```

Comment:
I don't think there is any need to refer to the Aussie Govt benchmark anymore. Anyway, it's fairly intuitive

Reply:
Removed, thanks!